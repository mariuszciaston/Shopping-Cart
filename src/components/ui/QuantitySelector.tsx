import { QuantitySelectorProps } from "@/types/types";
import { Button } from "./Button";
import { Input } from "./Input";

const QuantitySelector: React.FC<QuantitySelectorProps> = ({
  quantity,
  setQuantity,
}) => {
  const handleIncrement = () => {
    if (quantity < 99) {
      setQuantity(!isNaN(quantity) ? quantity + 1 : 1);
    }
  };

  const handleDecrement = () => {
    if (quantity > 1) {
      setQuantity(quantity - 1);
    }
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = parseInt(e.target.value);

    if (value < 1) {
      setQuantity(1);
    } else if (value > 99) {
      setQuantity(99);
    } else {
      setQuantity(!isNaN(value) ? value : 1);
    }
  };

  return (
    <div className="flex gap-2">
      <Button
        data-testid="quantity-lower"
        className="text-lg"
        variant="outline"
        onClick={handleDecrement}
      >
        -
      </Button>
      <Input
        data-testid="quantity-input"
        className="m-none w-[4rem] pr-0 text-center"
        type="number"
        min="1"
        max="99"
        value={quantity}
        onChange={handleInputChange}
      />
      <Button
        data-testid="quantity-boost"
        className="text-lg"
        variant="outline"
        onClick={handleIncrement}
      >
        +
      </Button>
    </div>
  );
};

export default QuantitySelector;
