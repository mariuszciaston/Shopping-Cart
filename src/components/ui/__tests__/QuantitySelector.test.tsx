import { render, screen, fireEvent } from "@testing-library/react";
import QuantitySelector from "../QuantitySelector";
import { describe, it, expect, vi } from "vitest";

describe("QuantitySelector", () => {
  let quantity = 1;
  const setQuantity = vi.fn((newQuantity) => {
    quantity = newQuantity;
  });

  const renderComponent = (quantity: number) => {
    return render(
      <QuantitySelector quantity={quantity} setQuantity={setQuantity} />,
    );
  };

  it("should render the QuantitySelector component correctly", () => {
    renderComponent(quantity);
    expect(screen.getByTestId("quantity-lower")).toBeInTheDocument();
    expect(screen.getByTestId("quantity-input")).toBeInTheDocument();
    expect(screen.getByTestId("quantity-boost")).toBeInTheDocument();
  });

  it("should increment the quantity when increment button is clicked", () => {
    renderComponent(quantity);
    const incrementButton = screen.getByTestId("quantity-boost");
    fireEvent.click(incrementButton);
    expect(setQuantity).toHaveBeenCalledWith(2);
  });

  it("should decrement the quantity when decrement button is clicked", () => {
    quantity = 2;
    renderComponent(quantity);
    const decrementButton = screen.getByTestId("quantity-lower");
    fireEvent.click(decrementButton);
    expect(setQuantity).toHaveBeenCalledWith(1);
  });

  it("should not decrement the quantity below 1", () => {
    renderComponent(quantity);
    const decrementButton = screen.getByTestId("quantity-lower");
    fireEvent.click(decrementButton);
    expect(setQuantity).toHaveBeenCalledWith(1);
  });

  it("should set the quantity to input value", () => {
    renderComponent(quantity);
    const input = screen.getByTestId("quantity-input");
    fireEvent.change(input, { target: { value: "5" } });
    expect(setQuantity).toHaveBeenCalledWith(5);
  });

  it("should not set the quantity above 99", () => {
    renderComponent(quantity);
    const input = screen.getByTestId("quantity-input");
    fireEvent.change(input, { target: { value: "100" } });
    expect(setQuantity).toHaveBeenCalledWith(99);
  });

  it("should not set the quantity below 1", () => {
    renderComponent(quantity);
    const input = screen.getByTestId("quantity-input");
    fireEvent.change(input, { target: { value: "0" } });
    expect(setQuantity).toHaveBeenCalledWith(1);
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <QuantitySelector quantity={quantity} setQuantity={setQuantity} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
