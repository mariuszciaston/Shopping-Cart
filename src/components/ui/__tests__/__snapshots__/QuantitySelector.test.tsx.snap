// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuantitySelector > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
      data-testid="quantity-lower"
    >
      -
    </button>
    <input
      class="flex h-9 rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm m-none w-[4rem] pr-0 text-center"
      data-testid="quantity-input"
      max="99"
      min="1"
      type="number"
      value="1"
    />
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
      data-testid="quantity-boost"
    >
      +
    </button>
  </div>
</DocumentFragment>
`;
